File: python/src/runapi/grok_imagine/resources/edit_image.py
"""Grok-Imagine prompt-guided image editing resource."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource, ValidationError

from ..contract_gen import CONTRACT
from ..types import (
    CompletedImageTaskResponse,
    ImageTaskResponse,
)


class EditImage(Resource):
    """Edit an image with a text prompt using Grok-Imagine."""

    ENDPOINT = "/api/v1/grok_imagine/edit_image"

    RESPONSE_CLASS = ImageTaskResponse
    COMPLETED_RESPONSE_CLASS = CompletedImageTaskResponse

    def run(self, **params: Any) -> Any:
        """Create an edit-image task and poll until it completes.

        Args:
            **params: Edit-image parameters (model, prompt, ...).

        Returns:
            The completed edit-image response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create an edit-image task and return immediately with an id.

        Args:
            **params: Edit-image parameters (model, prompt, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of an edit-image task.

        Args:
            id: Task id.

        Returns:
            The current edit-image status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._validate_contract(CONTRACT["edit-image"], params)
        if not params.get("source_image_url"):
            raise ValidationError("source_image_url is required")

File: ruby/lib/runapi/grok_imagine/resources/edit_image.rb
# frozen_string_literal: true

module RunApi
  module GrokImagine
    module Resources
      # Prompt-guided image editing resource.
      #
      # Image 2.0 accepts one to five source_image_urls, an aspect_ratio, and
      # an optional prompt. The original model continues to accept a single
      # source_image_url and prompt.
      class EditImage
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/grok_imagine/edit_image"

        RESPONSE_CLASS = Types::ImageTaskResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedImageTaskResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["edit-image"], params)
        end
      end
    end
  end
end
